src/main/webapp/location/location.tsx

```diff
@@ -0,0 +1,105 @@
+import * as React from 'react'
+import FormControl from '@material-ui/core/FormControl'
+import InputLabel from '@material-ui/core/InputLabel'
+import MenuItem from '@material-ui/core/MenuItem'
+import Select from '@material-ui/core/Select'
+import Line from './line'
+import Polygon from './polygon'
+import PointRadius from './point-radius'
+import BBox from './bbox'
+import withCoordinateUnitTabs from './with-coordinate-unit-tabs'
+import Keyword from './keyword'
+import { BasicEditorProps } from './geo-editor'
+import { geometry, shapes } from 'geospatialdraw'
+
+type LocationType = 'line' | 'polygon' | 'pointRadius' | 'bbox' | 'keyword'
+const defaultLocationType: LocationType = 'line'
+
+type LocationTypeComponentMap = {
+  [type in LocationType]: {
+    label: string
+    Component: React.ComponentType<BasicEditorProps>
+    shape: shapes.Shape
+  }
+}
+
+type EditorPropsMap = { [editorType in LocationType]: any }
+
+type Props = BasicEditorProps & {
+  editorProps?: EditorPropsMap
+}
+
+const componentMap: LocationTypeComponentMap = {
+  line: {
+    label: 'Line',
+    Component: withCoordinateUnitTabs(Line),
+    shape: shapes.LINE,
+  },
+  polygon: {
+    label: 'Polygon',
+    Component: withCoordinateUnitTabs(Polygon),
+    shape: shapes.POLYGON,
+  },
+  pointRadius: {
+    label: 'Point-Radius',
+    Component: withCoordinateUnitTabs(PointRadius),
+    shape: shapes.POINT_RADIUS,
+  },
+  bbox: {
+    label: 'Bounding Box',
+    Component: withCoordinateUnitTabs(BBox),
+    shape: shapes.BOUNDING_BOX,
+  },
+  keyword: {
+    label: 'Keyword',
+    Component: Keyword,
+    shape: shapes.POLYGON,
+  },
+}
+
+const Location: React.SFC<Props> = ({
+  value,
+  onChange,
+  editorProps = {} as EditorPropsMap,
+}) => {
+  const [locationType, setLocationType] = React.useState<LocationType>(
```

Comment:
If a component is taking in an onChange method, it shouldn't be keeping track of the state its calling the onChange method with. This data should be passed down as props.